### DISCUSSION TITLE
Predicting Preview URLs on Vercel

### DISCUSSION THREAD

**Discussion by leo-petrucci** [vercel/vercel#6045] [Help] [answered]
I'm quite confused by this so I hope someone can shed some light over how to accomplish this.

I currently have a serverless API which uses some OAuth login functionality, this means that I need to provide it a `callbackURL` when the external login has been completed, so Twitter/Facebook/Google can redirect the user back to my API to finish the login process.

This works great when my URL is static in production as I know where the API route is going to be on my domain, but when doing preview deployments my URLs are completely random, so I'm unsure how I could possibly predict how my URL will look and pass it as an Environment Variable?

I know my callback will be located at `/auth/twitter/callback`, but I'm having trouble figuring out how to predict what `https://serverless-pri-git-feature-reauth-header-cre-4221eb.vercel.app` will look like depending on the branch I'm previewing.

**Comment by mathieutu** [comment]
Linked issue:
auth0/nextjs-auth0#440
auth0/nextjs-auth0#417
https://github.com/vercel/vercel/discussions/6656

**Comment by leo** [comment, marked as answer]
Whenever you create a Deployment, Vercel will assign – what we call – "Automatic URLs". One of those is random, but some others exist that are not, so those are the ones you could rely on for this.

The ones that aren't random aren't exposed to your Deployment through Environment Variables yet (we're working on that), but you can construct them yourself, as they're based on pieces of information that you already have available ahead of time.

You can learn more about Automatic URLs on Vercel [here](https://vercel.com/docs/deployments/automatic-urls).

I also recommend taking a look at the "Preview Deployment Suffix" section in the documentation page linked above, as that feature lets you replace the `vercel.app` part of the URL with a Custom Domain, so if your third-party service supports wildcard callback URLs, you could define one like that (e.g. `*.my-domain.com`).

**Reply by ramiel** [reply]
For me, sometimes, all the produced urls contain random bits. For example the predictable url of this deployment is
`https://mysite-git-branchname-43039c-myorg.vercel.app` where `43039c` is not predictable ahead of time. This happens only sometimes, not always, but makes building url impossible

**Reply by leerob** [reply]
Did you have a chance to read the docs above? It mentions this

```
<project-name>-<unique-hash>-<scope-slug>.vercel.app
```

Let me know if something wasn't clear 😄

**Reply by ramiel** [reply]
Yes indeed, these are the deployment url. I'm talking about the [branch url](https://vercel.com/docs/deployments/automatic-urls#automatic-branch-urls) `<project-name>-git-<branch-name>-<scope-slug>.vercel.app`.     
From the documentation I understand there should be no random parts. Am I wrong?

**Reply by leerob** [reply]
Correct, branch URLs are structured differently. You are likely looking at an individual deployment URL, not a branch URL. Could you share an example?

**Reply by ramiel** [reply]
In the screenshot below you can see how some branch deployments have those random bits at the end. It happens only sometimes (last one is correct) but often and often

![Screenshot_20210913_152022](https://user-images.githubusercontent.com/324687/133091035-e199932e-9212-4598-88f2-5908dc92c126.png)

**Reply by mathieutu** [reply]
The thing here is that you provide us theses url in comments and as github branches preview url, BUT we can't get the url in our code via an environment variable, like we do with VERCEL_URL for deployment urls, so every app that needs this url in its backend (common need for oauth) is screwed.

**Reply by leo** [reply]
> For me, sometimes, all the produced urls contain random bits. For example the predictable url of this deployment is
> `https://mysite-git-branchname-43039c-myorg.vercel.app` where `43039c` is not predictable ahead of time. This happens only sometimes, not always, but makes building url impossible

Check out the "Truncation" section on [this documentation page](https://vercel.com/docs/deployments/automatic-urls#truncation) about Automatic URLs. We generally don't attach these random characters, only if your branch name is too long to fit into the Domain, since Domains have a maximum length in general (not just for our system on Vercel).

**Reply by leo** [reply]
We're looking to land a way in which the URL could be retrieved from a System Environment Variable.

At the moment, unfortunately, it's only known after the Deployment has become ready, so we can't yet easily expose it. Until the full fix has landed, I recommend constructing the URL as I've mentioned in my original reply (with branch names that aren't too long to fit into the URL), or using the [Preview Deployment Suffix](https://vercel.com/docs/deployments/automatic-urls#preview-deployment-suffix) feature with a Wildcard Domain in your third-party service.

We'll try to get the full fix landed as soon as possible! Thank you for your patience.

**Reply by ramiel** [reply]
Thank you @leo , this was the missing information. So it's the truncation that is applying here. I noticed that section in the documentation but I thought it was not applying to my case because `https://hse-homepage-git-952-hide-informations-about-users-hypersay-events.vercel.app` is only 85 characters and the branch name `952-hide-informations-about-users` is only 33 which are way less than what's on the docs (253 for the entire url and 63 for each part, so for the branch name). Again, probably I'm not really understanding the docs here

**Reply by ramiel** [reply]
Also, in case of monorepo, would be useful to have the branch name truncated with always the same characters, so it'd be easier to guess all the websites url starting from one. Maybe an SHA instead of random characters, or something similar?

**Reply by ramiel** [reply]
Oh, I see, where the documentation says `Furthermore, if a single component exceeds 63 characters, it will be truncated regardless of the above limit.` it doesn't refer to subdomain components (branch name, org scope and so on), but to each part of the domain `(63 letters).(63 letters).(63 letters).(62 letters)`. 

I think my confusion comes from the sentence above `If all the components that are used by a particular Automatic URL added together (including their separators like - or -git-)`, which seems to refer to parts of the subdomain.

**Reply by leo** [reply]
Oh, I see. Thanks for sharing that! We'll adjust the documentation to be more clear around this. I'll keep you in the loop!

**Reply by leo** [reply]
The documentation update is now available! Let me know if it's now clear.

**Reply by ramiel** [reply]
Awesome, thanks for clarifying things, it's perfect now.

**Reply by leo** [reply]
That's great to hear! Thanks for confirming.

**Reply by jozefhruska** [reply]
Unfortunately, it's not always possible to construct the Deployment URL from System Environment Variables exposed by Vercel. For example, if the branch name contains `/` (e.g. `fix/*`):
- Deployment URL: `https://<project-name>-git-fix-vercel-preview-environments-<scope-slug>.vercel.app`
- Exposed `VERCEL_GIT_COMMIT_REF` environment variable: `fix/vercel-preview-environments`

Although this seems like an easy transformation, we're using `.env.production` file to construct the URL (according to https://github.com/auth0/nextjs-auth0/tree/main/examples) and AFAIK it doesn't allow any transformations:
```
AUTH0_BASE_URL=${VERCEL_GIT_REPO_SLUG}-git-${VERCEL_GIT_COMMIT_REF}-${VERCEL_GIT_REPO_OWNER}.vercel.app
```
Bottom line is, it'd be really helpful if Vercel could expose the Git Deployment URL ☺️
